spec.md

```diff
@@ -3149,6 +3167,276 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+  This results in the CO receiving its language specific notification idiom for the end of a gRPC stream, which provides a **definitive indication that all available metadata has been received**.
+
+- If the SP encounters an error while recovering the metadata it MUST abort transmission of the stream with its language specific error idiom.
+  This results in the CO receiving the error in its language specific idiom, which will be different from the language specific idiom for the end of a gRPC stream.
+
+- It is possible that the gRPC stream gets interrupted for arbitrary reasons beyond the control of either the SP or the CO.
+  The SP will get an error when writing to the stream and MUST abort its transmission.
+  The CO will receive an error in its language specific idiom, which will be different from the language specific idiom for the end of a gRPC stream.
+
+- In all circumstances where the CO receives an error when reading from the gRPC stream it MAY attempt to **continue** the operation by re-sending its request message but with a `starting_offset` adjusted to the NEXT byte position beyond that of any metadata already received.
+  The SP MUST always ensure that the `starting_offset` requested be considered in the computation of the data range for the first message in the returned gRPC stream, though the data range of the first message is not required to actually include the `starting_offset` if there is no applicable data between the `starting_offset` and the start of the data range returned by the first message.
+
+#### `GetMetadataAllocated`
+
+The plugin must implement this RPC if it provides the SnapshotMetadata service.
+
+```protobuf
+// The GetMetadataAllocatedRequest message is used to solicit metadata
+// on the allocated blocks of a snapshot: i.e. this identifies the
+// data ranges that have valid data as they were the target of some
+// previous write operation on the volume.
+message GetMetadataAllocatedRequest {
+  // This is the identifier of the snapshot.
+  // This field is REQUIRED.
+  string snapshot_id = 1;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed.
+  // It is intended to be used to continue a previously interrupted
+  // call.
+  // The CO SHOULD specify this value to be the offset of the byte
+  // position immediately after the last byte of the last data range
+  // received, if continuing an interrupted operation, or zero if not.
+  // The SP MUST ensure that the returned response stream does not
+  // contain BlockMetadata tuples that start before but not overlap
+  // the requested starting_offset: i.e. if S is the requested
+  // starting_offset, and B0 is block_metadata[0] of the first message
+  // in the response stream, then the following must always be true:
+  // B0.byte_offset > S || S < B0.byte_offset + B0.size_bytes.
```

Comment:
```suggestion
  // S < B0.byte_offset or-else S < B0.byte_offset + B0.size_bytes.
```
^ this seems more readable? (for me, anyway); otherwise 👍

Reply:
No problems with switching the order of the first test, but "or-else" ???

How about if I change it to
> ...
> in the response stream, then either (S < B0.byte_offset) must
> be true or (S < B0.byte_offset + B0.size_bytes) must be true.